### PR TITLE
Movie Recommendation using VectorDB

### PR DESCRIPTION
**Pull request by vipul-maheshwari**
Article

### CODE REVIEW

**Comment by svonava-superlinked** [comment]
Hi @vipul-maheshwari thank you for contributing.

Before I dive into this, I noticed you say "production grade" in the summary - what for you makes this example qualify for "production grade"?

**Comment by vipul-maheshwari** [comment]
> Hi @vipul-maheshwari thank you for contributing.
> 
> Before I dive into this, I noticed you say "production grade" in the summary - what for you makes this example qualify for "production grade"?

Hey Hi @svonava-superlinked , Well the retrieval and the similarity checks with the genre embeddings makes it more robust in terms of the recommendation! If you think I need to change this, let me know!

File: docs/articles/movie_recommendation_using_vectordb.markdown [hunk 1]
```diff
@@ -0,0 +1,568 @@
...
+ 
+First, we initialize our LanceDB table containing our relevant data. Then, for a given movie title, we check if we have this title in our dataset. If it exists, we perform a cosine similarity search on all other movies, returning the top 5 most relevant movie titles.
+
+And just like that, you've crafted an impressive movie recommendation system that can swiftly provide you with the best selection of movies. This is made possible by leveraging `genre_embeddings` created with the assistance of a neural network and storing them in LanceDB vector database.
+
+Doing all this will provide you with a list of recommended movies.
+
+*Just to clarify, I've employed a rudimentary deep learning model architecture to showcase how we can improve our recommendation system. While the movie recommendations may not yield optimal results, the team at Tubi would be the experts to provide you with the exact model architecture they've utilized for developing this system.*
+
+That being said, Here's an overview of the entire training code..
```
Comment (morkapronczay) [inline docs/articles/movie_recommendation_using_vectordb.markdown:417]: I don't think the entire code is needed here if you link to a colab with the code, and parts are there in the text.

File: docs/articles/movie_recommendation_using_vectordb.markdown [hunk 2]
```diff
@@ -0,0 +1,568 @@
...
+movie_data = pd.read_csv('data/movies_metadata.csv', low_memory=False)
+device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+
+def preprocess_data(movie_data_chunk):
+    tagged_docs = []
+    valid_indices = []
+    movie_info = []
+
+    # Wrap your loop with tqdm
+    for i, row in tqdm(movie_data_chunk.iterrows(), total=len(movie_data_chunk)):
```
Comment (morkapronczay) [inline docs/articles/movie_recommendation_using_vectordb.markdown:445]: iterrows is probably the least efficient way to work through a dataframe:
check the second answer here: https://stackoverflow.com/questions/16476924/how-can-i-iterate-over-rows-in-a-pandas-dataframe

**Review by morkapronczay — CHANGES_REQUESTED** [review]
Thank you for your article, please read all my notes as I am trying to help! I really appreciate the effort that went into this work!

I like the approach of:
1. Creating a custom doc2vec model (trained on title + genre concatenated in a structured way)
2. Train a neural network on a relevant task using the doc2vec vectors as input
3. Extract the embeddings from a hidden layer of the neural network
4. Use those representations in a vector database to produce recommendations.

My concerns are the following:
- The doc2vec model is trained on a very specific text structure (title + genre structured),  then at inference you use it to embed movie descriptions. This makes it questionable for me that this model would withstand a more rigorous evaluation. Even the 3 provided examples are not that convincing.
- The whole model is trained to predict genre, therefore the stored representations of movies are useful for that purpose too. I believe as a result, the recommender system will essentially just return similar genre movies.

The way I see it these last 2 points makes it questionable if this is approach worth promoting.

Some steps I would see as beneficial:
- Structure the testing movie descriptions in a similar way as you do with your training data
- Use movie descriptions too, when training the doc2vec model (and maybe simply use a sentence_transformers model instead of the doc2vec)
- consider training the model for a different task than genre prediction - ideal would be movie pairs that are frequently watched together, for example

I am open to discuss these points, if you need me you can find on the VectorHub Slack!

**Comment by vipul-maheshwari** [comment]
Thanks for the requested changes @morkapronczay , I am going to update the branch as per the requested changes and feedback. Pinging you soon

File: docs/articles/movie_recommendation_using_vectordb.markdown [hunk 3]
```diff
@@ -0,0 +1,568 @@
+# Movie Recommendation using VectorDB
+
+This article provides a comprehensive guide on creating a scalable and production-grade movie recommendation system by leveraging the power of Genre Spectrum Embeddings and VectorDB. Inspired by the implementation outlined in this [paper](https://arxiv.org/pdf/2309.08787), We'll explore how combining these two techniques can significantly enhance the recommendation experience, addressing key challenges faced by traditional systems.
```
[the remark this answers is not in the document]
Comment (vipul-maheshwari) [reply docs/articles/movie_recommendation_using_vectordb.markdown:3]: Vector databases in general! Changing the VectorDB to vector databases to make it more robust!

File: docs/articles/movie_recommendation_using_vectordb.markdown [hunk 4]
```diff
@@ -0,0 +1,568 @@
...
+
+Let's get started!
+
+## Introduction to Genre Spectrum Embeddings
+
+Scrolling through streaming platforms can be frustrating when the movie suggestions don't match our interests. Building recommendation systems is a complex task as there isn't one metric that can measure the quality of recommendations. To improve this, we propose combining Genre Spectrum Embeddings and VectorDB for better recommendations.
+
+The Genre Spectrum approach involves combining the various movie genres or characteristics of a movie to form Initial embeddings, which offer a comprehensive portrayal of the movie content. Then these embeddings are used as a input to train a Deep Learning model producing Genre Spectrum embeddings at the penultimate layer. 
+
+These embeddings serve dual purposes: they can either be directly inputted into a classification model for genre classification or stored in a VectorDB. By storing embeddings in a VectorDB, efficient retrieval and query search for recommendations become possible at a later stage. This architecture offers a holistic understanding of the underlying processes involved.
```
[the remark this answers is not in the document]
Comment (vipul-maheshwari) [reply docs/articles/movie_recommendation_using_vectordb.markdown:24]: Done! Changed!

**Comment by vipul-maheshwari** [comment]
### Changes to the Article Outline

Hello @morkapronczay,

I am making some changes to the article outline for a few reasons:

1. **Using embeddings from the hidden layer**: Initially, we planned to use embeddings from the hidden layer for vector search. However, this approach seems too advanced given our basic neural network with only a couple of layers. The results may not be as effective as we hoped.

2. **New approach**: Instead of using embeddings, I will use the classified genres. I have already implemented pre-filtering with LanceDB. Combining the classified genres and pre-filtering will help us provide better recommendations.

### Architectural Overview

Here is the updated architectural overview:

- **Classified Genres**: Use the neural network to classify genres.
- **Pre-filtering with LanceDB**: Use LanceDB to pre-filter the classified genres.
- **Combine Both**: Combine the results from the classified genres and pre-filtering to give recommendations.

This new approach is more effective for our current setup.

File: docs/articles/movie_recommendation_using_vectordb.markdown (hunk 1 above)
Comment (vipul-maheshwari) [reply docs/articles/movie_recommendation_using_vectordb.markdown:417]: Done!

File: docs/articles/movie_recommendation_using_vectordb.markdown [hunk 5]
```diff
@@ -0,0 +1,568 @@
...
+    movie_data_chunk = movie_data.iloc[chunk_start:chunk_start+chunk_size]
+    chunk_tagged_data, chunk_valid_indices, chunk_movie_info = preprocess_data(movie_data_chunk)
+    tagged_data.extend(chunk_tagged_data)
+    valid_indices.extend(chunk_valid_indices)
+    movie_info.extend(chunk_movie_info)
+```
+
+## Generating embeddings using Doc2Vec
+
+Next, we'll utilize the Doc2Vec model to generate embeddings for each movie based on the preprocessed text. We'll allow the Doc2Vec model to train for several epochs to capture the essence of the various movies and their metadata in the multidimensional latent space. This process will help us represent each movie in a way that captures its unique characteristics and context.
```
[the remark this answers is not in the document]
Comment (vipul-maheshwari) [reply docs/articles/movie_recommendation_using_vectordb.markdown:100]: I wanted to showcase the raw setup of how embeddings work at the document or paragraph level.

[the remark this answers is not in the document]
Comment (vipul-maheshwari) [reply docs/articles/movie_recommendation_using_vectordb.markdown:100]: I think keeping it that way would make more sense to the folks who are new to embeddings and just wanted to understand how the things are working behind the scenes!

File: docs/articles/movie_recommendation_using_vectordb.markdown [hunk 6]
```diff
@@ -0,0 +1,568 @@
...
+7. Using Genre Spectrum Embeddings to get the relevant recommendations
+8. Get the relevant recommendation for a given Movie
+
+Let's get started!
+
+## Introduction to Genre Spectrum Embeddings
+
+Scrolling through streaming platforms can be frustrating when the movie suggestions don't match our interests. Building recommendation systems is a complex task as there isn't one metric that can measure the quality of recommendations. To improve this, we propose combining Genre Spectrum Embeddings and VectorDB for better recommendations.
+
+The Genre Spectrum approach involves combining the various movie genres or characteristics of a movie to form Initial embeddings, which offer a comprehensive portrayal of the movie content. Then these embeddings are used as a input to train a Deep Learning model producing Genre Spectrum embeddings at the penultimate layer.
```
[the remark this answers is not in the document]
Comment (vipul-maheshwari) [reply docs/articles/movie_recommendation_using_vectordb.markdown:22]: Done! Changed

File: docs/articles/movie_recommendation_using_vectordb.markdown [hunk 7]
```diff
@@ -0,0 +1,568 @@
...
+
+![image](https://github.com/vipul-maheshwari/vipul-maheshwari.github.io/blob/main/images/movie-recommendation-using-rag/movie_recommendation_architecture.png?raw=true)
+
+## Data Ingestion and preprocessing techniques for movie metadata
+
+Our initial task involves gathering and organizing information about movies. This includes gathering extensive details such as the movie's type, plot summary, genres, audience ratings, and more.
+
+Thankfully, we have access to a robust dataset on [Kaggle](https://www.kaggle.com/datasets/rounakbanik/the-movies-dataset) containing information from various sources for approximately 45,000 movies. Please dowload the data from the Kaggle and place it inside your working directory. Watch for a file named `movies_metadata.csv`. If you require additional data, you can supplement the dataset by extracting information from platforms like Rotten Tomatoes, IMDb, or even box-office records.
```
[the remark this answers is not in the document]
Comment (vipul-maheshwari) [reply docs/articles/movie_recommendation_using_vectordb.markdown:32]: Done! Changed

File: docs/articles/movie_recommendation_using_vectordb.markdown [hunk 8]
```diff
@@ -0,0 +1,568 @@
...
+    for epoch in range(num_epochs):
+        doc2vec_model.train(tqdm(tagged_data, desc=f"Epoch {epoch+1}"), total_examples=doc2vec_model.corpus_count, epochs=doc2vec_model.epochs)
+    
+    return doc2vec_model
+
+doc2vec_model = train_doc2vec_model(tagged_data)
+doc2vec_model.save("doc2vec_model")
+```
+
+The `train_doc2vec_model` function trains a Doc2Vec model on the tagged movie data, producing 100-dimensional embeddings for each movie. These embeddings act as input features for the subsequent neural network, which we'll employ to generate the `genre_embeddings`.
```
[the remark this answers is not in the document]
Comment (vipul-maheshwari) [reply docs/articles/movie_recommendation_using_vectordb.markdown:115]: Done! Changed the `genre_embeddings` to the `embeddings` refer to my latest comment on why I am changing the article outline!

File: docs/articles/movie_recommendation_using_vectordb.markdown (hunk 2 above)
Comment (vipul-maheshwari) [reply docs/articles/movie_recommendation_using_vectordb.markdown:445]: Oh thanks @morkapronczay, I am changing the `iterrows` to something more efficient

Comment (vipul-maheshwari) [reply docs/articles/movie_recommendation_using_vectordb.markdown]: Thanks, I updated with the list comprehensions and it's damn fast!

File: docs/articles/movie_recommendation_using_vectordb.markdown [hunk 9]
```diff
@@ -0,0 +1,568 @@
...
+            tagged_docs.append(TaggedDocument(words=word_tokenize(movies_text.lower()), tags=[str(i)]))
+            valid_indices.append(i)
+            movie_info.append((row['title'], genres))
+        except Exception as e:
+            continue
+
+    return tagged_docs, valid_indices, movie_info
+
+# Preprocess data and extract genres for the first 1000 movies
+chunk_size = 1000
```
[the remark this answers is not in the document]
Comment (vipul-maheshwari) [reply docs/articles/movie_recommendation_using_vectordb.markdown]: Done!

File: docs/articles/movie_recommendation_using_vectordb.markdown [hunk 10]
```diff
@@ -0,0 +1,568 @@
...
+import torch.nn as nn
+import torch.optim as optim
+from nltk import word_tokenize
+from torch.utils.data import DataLoader, TensorDataset
+from gensim.models.doc2vec import Doc2Vec, TaggedDocument
+from sklearn.preprocessing import MultiLabelBinarizer
+from sklearn.model_selection import train_test_split
+from tqdm import tqdm
+
+import nltk
```
[the remark this answers is not in the document]
Comment (vipul-maheshwari) [reply docs/articles/movie_recommendation_using_vectordb.markdown:50]: Edit:  Spacy was taking too much time as compared to nltk, and the accuracy dropped!
Going with nltk for now!

File: docs/articles/movie_recommendation_using_vectordb.markdown [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+        .limit(10)
+        .where(where_clause)
+        .to_pandas()
+    )
+    return result[["title"]]
+
+get_recommendations("Toy Story")
+```
+
+![results]()
```
Comment (morkapronczay) [inline docs/articles/movie_recommendation_using_vectordb.markdown:343]: missing image

File: docs/articles/movie_recommendation_using_vectordb.markdown [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+
+Let's get started!
+
+## Why use embeddings for recommendation systems?
+
+Scrolling through streaming platforms can be frustrating when the movie suggestions don't match our interests. Building recommendation systems is a complex task, as there isn't one metric that can measure the quality of recommendations. To improve this, we can combine embeddings and vector database for better recommendations. These embeddings serve dual purposes: they can either be directly used as input to a classification model for genre classification or stored in a vector database for retrieval purposes. By storing embeddings in a vector database, efficient retrieval and query search for recommendations become possible at a later stage.
+
+This architecture offers a holistic understanding of the underlying processes involved.
+
+![image]()
```
Comment (morkapronczay) [inline docs/articles/movie_recommendation_using_vectordb.markdown:20]: missing image

**Review by morkapronczay — COMMENTED** [review]
Thanks for the fixes!

I think there are 2 more things left to do:
- correct the missing images
- put a paragraph at the beginning that describes how many different type of data can be embedded, but for the sake of simplicity you will create embeddings based on title and genre data - therefore the recommendations are also based on similar genres and titles

File: docs/articles/movie_recommendation_using_vectordb.markdown [hunk 13]
```diff
@@ -0,0 +1,347 @@
...
+# Saving the trained model
+torch.save(model.state_dict(), 'trained_model.pth')
+```
+
+## A movie recommendation system
+
+To create our movie recommendation system, we'll implement a process where users input a movie title, and we generate relevant recommendations based on that input. Key to this approach is utilizing Doc2Vec embeddings, which we'll store in a vector database for efficient retrieval.
+
+Here's how it works: When a user enters a movie title as a query, we'll first retrieve its embeddings from our vector database. With these embeddings in hand, our next step involves identifying 'n' number of movies whose embeddings closely resemble those of our query movie. We can achieve this by employing similarity metrics such as cosine similarity or by determining the smallest Euclidean distances. By leveraging these techniques, our system will provide users with personalized movie recommendations that are closely aligned with their initial movie interests. This approach ensures an intuitive and effective recommendation process tailored to individual preferences based on the genres.
+
```
Comment (robertdhayanturner) [inline docs/articles/movie_recommendation_using_vectordb.markdown:274]: @vipul-maheshwari 
according to your discussion, all we have from the user is their query. Here for the first time you introduce "initial movie interests" and "individual preferences based on genres" - where do these come from? They're not in your code or discussion.

File: docs/articles/movie_recommendation_using_vectordb.markdown [hunk 14]
```diff
@@ -0,0 +1,347 @@
...
+# preproces the documents, and create TaggedDocuments
+tagged_data = [TaggedDocument(words=word_tokenize(doc.lower()),
+                              tags=[str(i)]) for i,
+               doc in enumerate(complete_data)]
+```
+
+## Generating embeddings using Doc2Vec
+
+Next, we'll utilize the Doc2Vec model to generate embeddings for each movie based on the preprocessed text. We'll allow the Doc2Vec model to train for several epochs to capture the essence of the various movies and their metadata in the multidimensional latent space. This process will help us represent each movie in a way that captures its unique characteristics and context.
+
```
Comment (robertdhayanturner) [inline docs/articles/movie_recommendation_using_vectordb.markdown:107]: @vipul-maheshwari 
the "preprocessed text" and "metadata" here are genre and text only, correct?

File: docs/articles/movie_recommendation_using_vectordb.markdown (hunk 13 above)
Comment (vipul-maheshwari) [reply docs/articles/movie_recommendation_using_vectordb.markdown:274]: Yes! So let's say If a user searches for 'Toy Story,' it indicates their interest in movies that fall under the genres of "Animation" or "Comedy"

File: docs/articles/movie_recommendation_using_vectordb.markdown (hunk 14 above)
Comment (vipul-maheshwari) [reply docs/articles/movie_recommendation_using_vectordb.markdown:108]: Yes! Genres and the movie Titles!

**Comment by vipul-maheshwari** [comment]
Looking goooood! 🚀
Thanks @robertdhayanturner @morkapronczay , we did this together! More to come
